File: leetcode/438.FindAllAnagramsInAString.go
package leetcode

import "reflect"

func findAnagrams(s string, p string) []int {
	if len(p) > len(s) {
		return []int{}
	}

	sCount := make(map[string]int, len(s))
	pCount := make(map[string]int, len(p))

	for i := 0; i < len(p); i++ {
		if _, ok := pCount[string(p[i])]; !ok {
			pCount[string(p[i])] = 0
		}
		pCount[string(p[i])]++

		if _, ok := sCount[string(s[i])]; !ok {
			sCount[string(s[i])] = 0
		}
		sCount[string(s[i])]++
	}

	var result []int
	if reflect.DeepEqual(pCount, sCount) {
		result = append(result, 0)
	}
	left := 0
	for right := len(p); right < len(s); right++ {
		if _, ok := sCount[string(s[right])]; !ok {
			sCount[string(s[right])] = 0
		}
		sCount[string(s[right])]++
		sCount[string(s[left])]--

		if sCount[string(s[left])] == 0 {
			delete(sCount, string(s[left]))
		}
		left++

		if reflect.DeepEqual(sCount, pCount) {
			result = append(result, left)
		}
	}

	return result
}

File: leetcode/438.FindAllAnagramsInAString.ts
function findAnagrams(s: string, p: string): number[] {
  if (p.length > s.length) return []

  const pCount: Record<string, number> = {}
  const sCount: Record<string, number> = {}

  for (let i = 0; i < p.length; i++) {
    pCount[p[i]] = (pCount[p[i]] ?? 0) + 1
    sCount[s[i]] = (sCount[s[i]] ?? 0) + 1
  }

  const isEqual = (obj1: any, obj2: any): boolean => {
    for (const element in obj1) {
      if (!Object.prototype.hasOwnProperty.call(obj2, element)) return false
      if (obj2[element] !== obj1[element]) return false
    }
    return true
  }

  const result = isEqual(pCount, sCount) ? [0] : []
  let left = 0
  for (let right = p.length; right < s.length; right++) {
    sCount[s[right]] = (sCount[s[right]] ?? 0) + 1
    sCount[s[left]]--

    if (sCount[s[left]] === 0) {
      delete sCount[s[left]]
    }
    left++

    if (isEqual(sCount, pCount)) result.push(left)
  }

  return result
}
console.log(findAnagrams('cbaebabacd', 'abc'))
